From: NeilBrown <neilb@suse.de>
Subject: Allow read error in single drive in raid1 to be passed up
Patch-mainline: 2.6.29
References: bnc#447835

If a raid1 only has a single working device and gets a read error, 
we choose to simply return that error up to the filesystem (or whatever)
rather than failing the whol array.

However the codes doesn't quite do that.  We attempt a readbalance
which allocated the same drive, so we retry the read - indefinitely. 

Instead:  If read_balance in the error case choose the same drive that just
failed, treat it as a failure and don't retry.

Signed-off-by: Neil Brown <neilb@suse.de>

---
 drivers/md/raid1.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/md/raid1.c
+++ b/drivers/md/raid1.c
@@ -1637,7 +1637,8 @@ static void raid1d(mddev_t *mddev)
 			}
 
 			bio = r1_bio->bios[r1_bio->read_disk];
-			if ((disk=read_balance(conf, r1_bio)) == -1) {
+			if ((disk=read_balance(conf, r1_bio)) == -1 ||
+			    disk == r1_bio->read_disk) {
 				printk(KERN_ALERT "raid1: %s: unrecoverable I/O"
 				       " read error for block %llu\n",
 				       bdevname(bio->bi_bdev,b),
